scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/synchronization/AOSynchronization.java

```diff
@@ -0,0 +1,835 @@
+
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    @Override
+    public void conditionalReplaceAll(String originator, TaskId taskid, String channel, String predicate,
+            String thenRemappingFunction, String elseRemappingFunction)
+            throws InvalidChannelException, CompilationException, IOException {
+        String currentKey = null;
+        try {
+            Channel<Serializable> chosenChannel = getChannel(channel);
+            for (Map.Entry<String, Serializable> entry : chosenChannel.entrySet()) {
+                currentKey = entry.getKey();
+                Serializable value = chosenChannel.get(currentKey);
+                if (evaluateClosure(predicate, BiPredicate.class).test(currentKey, value)) {
+
+                    Serializable answer = chosenChannel.compute(currentKey,
+                                                                (BiFunction<? super String, ? super Serializable, ? extends Serializable>) evaluateClosure(thenRemappingFunction,
+                                                                                                                                                           BiFunction.class));
+                    logWithContext(taskid,
+                                   channel,
+                                   "Run conditionalReplaceAll(Then)" + ON_KEY + QUOTE + currentKey + QUOTE + RESULT_IS +
+                                            answer,
+                                   Level.DEBUG);
+                } else {
+                    Serializable answer = chosenChannel.compute(currentKey,
+                                                                (BiFunction<? super String, ? super Serializable, ? extends Serializable>) evaluateClosure(elseRemappingFunction,
+                                                                                                                                                           BiFunction.class));
+                    logWithContext(taskid,
+                                   channel,
+                                   "Run conditionalReplaceAll(Else)" + ON_KEY + QUOTE + currentKey + QUOTE + RESULT_IS +
+                                            answer,
+                                   Level.DEBUG);
+
+                }
+            }
+            commitIfNeeded(channel);
+
+        } catch (InvalidChannelException | CompilationException | IOException e) {
+            throw e;
+        } catch (Throwable e) {
+            throw new ClosureEvaluationException(EXCEPTION_WHEN_EVALUATING_CLOSURE + StackTraceUtil.getStackTrace(e));
+        }
+    }
+
+    @Override
+    public boolean waitUntil(String originator, TaskId taskid, String channel, String key, String predicate)
```

Comment:
I do not see where in the code we actually do: `while (!predicate.test(key, oldValue))`

Reply:
This part is done inside the runActivity method, a complex ActiveObject magic.